lib/pos_mempool.go

```diff
@@ -277,6 +277,94 @@ func (mp *PosMempool) IsRunning() bool {
 	return mp.status == PosMempoolStatusRunning
 }
 
+// OnBlockConnected is an event handler provided by the PoS mempool to handle the blockchain
+// event where a block is connected to the tip of the blockchain. The mempool updates its
+// internal state based on the new block that has been connected.
+//
+// Whenever a block is connected, this event handler removes the block's transactions from
+// the mempool and updates the internal fee estimation to include new block.
+func (mp *PosMempool) OnBlockConnected(block *MsgDeSoBlock) {
+	mp.Lock()
+	defer mp.Unlock()
+
+	if block.Header == nil || !mp.IsRunning() {
+		return
+	}
+
+	// Remove all transactions in the block from the mempool.
+	for _, txn := range block.Txns {
+		txnHash := txn.Hash()
+
+		// This should never happen. We perform a nil check on the txn hash to avoid a panic.
+		if txnHash == nil {
+			continue
+		}
+
+		// Get the transaction from the register. If the txn doesn't exist in the register,
+		// then there's nothing left to do.
+		txn := mp.txnRegister.GetTransaction(txnHash)
+		if txn == nil {
+			continue
+		}
+
+		mp.removeTransactionNoLock(txn, true)
+	}
+
+	// Add the block to the fee estimator. This is a best effort operation. If we fail to add the block
+	// to the fee estimator, we log an error and continue.
+	if err := mp.feeEstimator.AddBlock(block); err != nil {
+		glog.Errorf("PosMempool.OnBlockConnected: Problem adding block to fee estimator: %v", err)
+	}
+}
+
+// OnBlockDisconnected is an event handler provided by the PoS mempool to handle the blockchain
+// event where a block is disconnected from the tip of the blockchain. The mempool updates its
+// internal state based on the new block that has been connected.
+//
+// Whenever a block is disconnected, this event handler adds the block's transactions back to
+// the mempool and updates the internal fee estimation to exclude the disconnected block.
+func (mp *PosMempool) OnBlockDisconnected(block *MsgDeSoBlock) {
+	mp.Lock()
+	defer mp.Unlock()
+
+	if block.Header == nil || !mp.IsRunning() {
+		return
+	}
+
+	// Remove all transactions in the block from the mempool.
+	for _, txn := range block.Txns {
+		txnHash := txn.Hash()
+
+		// This should never happen. We perform a nil check on the txn hash to avoid a panic.
+		if txnHash == nil {
+			continue
+		}
+
+		// Add all transactions in the block to the mempool.
+
+		// Construct the MempoolTx from the MsgDeSoTxn.
+		mempoolTx, err := NewMempoolTx(txn, block.Header.TstampNanoSecs/1000, mp.latestBlockHeight)
```

Comment:
Interesting, there's no way for us to know exactly when this txn was actually added. This seems like the best approach since we know it was added to some mempool in the network before this timestamp.

Reply:
No need. This solution ends up working nicely. Assume we have txn A from block 1 and txn B from block 2, and block 2 has a higher timestamp than block 1. If we disconnect both blocks 1 and 2, then txn A will have precedence over txn B. And all txns in the mempool that didn't end up in blocks 1 or 2 must have not been proposed in the blocks because they either have lower fee buckets or higher timestamps.

Reply:
I think this is fine. There may be an extreme edge case where transaction reach certain nodes before a block is produced by another node, but I think it all washes out in the end. I tried to type it out multiple times and haven't been able to concretely define a case where this approach would break, so I think we're okay here.